common/src/character_controller.rs

```diff
@@ -35,11 +38,29 @@ impl<T> CharacterControllerPass<'_, T> {
         let movement = sanitize_motion_input(self.input.movement);
 
         if self.input.no_clip {
+            // If no-clip is on, the velocity field is useless, and we don't want to accidentally
+            // save velocity from when no-clip was off.
+            *self.velocity = na::Vector3::zeros();
             self.position.local *= math::translate_along(
                 &(movement * self.cfg.no_clip_movement_speed * self.dt_seconds),
             );
         } else {
-            // Placeholder: If no-clip is off, the character cannot move.
+            let old_velocity = *self.velocity;
+
+            // Update velocity
+            let current_to_target_velocity = movement * self.cfg.max_ground_speed - *self.velocity;
+            let max_delta_velocity = self.cfg.ground_acceleration * self.dt_seconds;
+            if current_to_target_velocity.norm_squared() > max_delta_velocity.powi(2) {
+                *self.velocity += current_to_target_velocity.normalize() * max_delta_velocity;
+            } else {
+                *self.velocity += current_to_target_velocity;
+            }
+
+            // Update position using the average between the old and new velocity to avoid the following two issues:
+            // 1. Input lag, which would occur if only the old velocity was used
+            // 2. Discontinuities, which would occur if only the new velocity was used
+            self.position.local *=
+                math::translate_along(&((*self.velocity + old_velocity) * 0.5 * self.dt_seconds));
```

Comment:
I found this difficult to follow. Is the idea that, by averaging the velocities together, we're actually modeling a velocity that increases linearly over the timestep? That makes sense, but isn't clearly communicated by the comment. In particular, I don't see how using the new velocity directly would be strictly more discontinuous, since it might model a temporary spike in velocity by the same token that this models a linear ramp to the end velocity.

Reply:
Your wording of treating velocity as increasing linearly over the timestep makes more sense than my wording, and I believe that is the general idea.

As for discontinuities, I made a mistake, as I had encountered what felt like discontinuities in the past, but that's unrelated to the code here and was due to a now-fixed bug. The main issue I encounter now with only using the new velocity is that when you let go of the "forward" button, you go backwards a little bit at the end. I assume you also accelerate more quickly at the beginning, but that's harder to notice.

I'll update the comment.